### ISSUE TITLE
setMonitor confusion

### ISSUE THREAD

**Issue by keshto** [LBCrion/sfwbar#84]
I have:
```
function("SfwbarInit") {
  SetLayer "top"
  SetMonitor "eDP-1"
}
```
which just plops the panel on whichever output the command is invoked on.
using `sfwbar -m eDP-1` works as expected in other words places the bar on `eDP-1` regardless of where it is invoked from.

**Comment by LBCrion** [comment]
Do you have a name assigned to the bar by any chance?
I.e. layout "mybar" {...}

**Comment by keshto** [comment]
nope just changing the default `t2.config` file
```
diff --git a/config/t2.config b/config/t2.config
index e002c76..4d54b9a 100644
--- a/config/t2.config
+++ b/config/t2.config
@@ -16,7 +16,7 @@ switcher {
 }
 
 function("SfwbarInit") {
-  SetBarId "bar-0"
+  SetMonitor "eDP-1"
   SetLayer "top"
 }
'''

**Comment by LBCrion** [comment]
Can you try running sfwbar with ``-d -g action`` ?

**Comment by keshto** [comment]
yes, DP-1 is primary and active external monitor and yes to weird behavior.

output with correct:
```
◗ sfwbar -d -g action -f t2.config 
16:19:07.22 widget action: (392) eDP-1 (addr null)
16:19:07.22 widget action: (393) top (addr null)
```

with invalid:
```
◗ sfwbar -d -g action -f t2.config 
16:19:18.43 widget action: (392) eDP-11 (addr null)
16:19:18.44 widget action: (393) top (addr null)
```

with external monitor (which works):
```
◗ sfwbar -d -g action -f t2.config 
16:20:52.24 widget action: (392) DP-1 (addr null)
16:20:52.26 widget action: (393) top (addr null)
```


they all look the same to me :confused:

**Comment by LBCrion** [comment]
Ok, the action is called correctly it seems. You're running this from DP-1 and in the first case the bar stays on DP-1 and in case 2 and 3 it ends up on eDP-1?

**Comment by keshto** [comment]
yes running from `DP-1`. No, it ends up on `eDP-1` ONLY with the invalid config ie. `eDP-11` or non existent display output.

**Comment by LBCrion** [comment]
Ok, I managed to replicate this locally, but only on labwc. On sway, it behaves as expected. Let me do some debugging and see what's causing this issue.

**Comment by keshto** [comment]
Ah, should have let you known which compositor I was running. So, you think it is a labwc issue?

**Comment by Consolatis** [comment]
Might help to diagnose the issue:
```diff
diff --git a/src/layers.c b/src/layers.c
index bd0b0e1..f98ea13 100644
--- a/src/layers.c
+++ b/src/layers.c
@@ -353,6 +353,12 @@ handle_new_layer_surface(struct wl_listener *listener, void *data)
 			return;
 		}
 		layer_surface->output = output;
+		wlr_log(WLR_ERROR, "new layer surface: no output requested, "
+			"falling back to output %s (from cursor coords).",
+			layer_surface->output->name);
+	} else {
+		wlr_log(WLR_ERROR, "new layer surface: output %s requested",
+			layer_surface->output->name);
 	}
 
 	struct lab_layer_surface *surface = znew(*surface);
```

I can look into it later as well if preferred.

**Comment by LBCrion** [comment]
Ok, I tracked this down to detection of the initial monitor on which the bar launches. This is not a labwc issue.

**Comment by keshto** [comment]
Currently it is launching the bar to whichever output has focus ie where the bar is invoked from which I think is correct behavior. The issue is that when one specifies that they want it on a non primary output it does not go there.

For all the tests above I set my external monitor as the primary.

**Comment by keshto** [comment]
Currently it is launching the bar to whichever output has focus ie where the bar is invoked from which I think is correct behavior. The issue is that when one specifies that they want it on a non primary output it does not go there.

For all the tests above I had my external monitor set as the primary.

**Comment by LBCrion** [comment]
I think the definition of primary here is whichever output is listed first in GdkDisplay (i.e. if you do sfwbar -m list, the first output listed is the "primary" one).

@Consolatis, I'd be interested to hear your opinion here. From what I can see gtk never receives wl_surface.enter for a new layer shell surface.

I.e. Let's say we have 2 outputs, WL-1 and WL-2. They are enumerated in this order in GdkDisplay. An application is launched from WL-2 and opens a layer shell window. Using your (very helpful) code above, I see that the layer shell window is assigned to WL-2 output based on the current focus. The client doesn't seem to receive wl_surface.enter to know that the layer shell window is on WL-2, so when the application runs gdk_display_get_monitor_at_window, gdk goes into a fallback.

Gdk fallback is based on global coordinates: 1) window overlap with various outputs, 2) output containing the center of the window. Since in wayland the client doesn't know the coordinates of the window in the global coordinate space, gdk_display_get_monitor_at_window returns the first monitor on the list.

From what I can see the wl.enter event is the missing bit in this scenario, and it seems to be affecting multiple compositors, which makes me wonder if it's an omission in wlroots?

**Comment by Consolatis** [comment]
Just a shot in the dark, I had a similar issue with the foreign-toplevel output_enter/leave events before: The outputs have to be bound before the events come through. Maybe GTK maps the surface and binds the outputs later?

I tested with an older version of sfwbar and wlroots seems sending the events correctly:
```
test@odroidc4:/home/user/dev/labwc$ WAYLAND_DEBUG=1 /home/user/dev/sfwbar/build/sfwbar --config ~/.config/sfwbar/own_bottom.config -d 2>&1 | grep 'wl_surface.*enter('
[ 651595.956] wl_surface@33.enter(wl_output@22)
^C
```
I will test with a recent build of sfwbar later.

Small debug diff of wlroots 0.16.2 (but should apply to all of wlroots 0.16).
That one can be used to differentiate between sending the outputs immediately or waiting for a future output bind.
In case of labwc it should be enough to modify subprojects/wlroots/ and then force labwc to use the subproject over a system wide installation of wlroots: `meson configure build --force-fallback-for=wlroots`

```diff
diff --git a/types/wlr_compositor.c b/types/wlr_compositor.c
index fa4a8aa5..3573a2b3 100644
--- a/types/wlr_compositor.c
+++ b/types/wlr_compositor.c
@@ -873,6 +873,9 @@ static void surface_handle_output_bind(struct wl_listener *listener,
 	struct wl_client *client = wl_resource_get_client(
 			surface_output->surface->resource);
 	if (client == wl_resource_get_client(evt->resource)) {
+		wlr_log(WLR_ERROR,
+			"wl_surface_send_enter on surface_handle_output_bind: %s",
+			surface_output->output->name);
 		wl_surface_send_enter(surface_output->surface->resource, evt->resource);
 	}
 }
@@ -912,6 +915,8 @@ void wlr_surface_send_enter(struct wlr_surface *surface,
 
 	wl_resource_for_each(resource, &output->resources) {
 		if (client == wl_resource_get_client(resource)) {
+			wlr_log(WLR_ERROR, "wl_surface_send_enter: %s",
+				surface_output->output->name);
 			wl_surface_send_enter(surface->resource, resource);
 		}
 	}
```
</detaails>

which, in my case with the older sfwbar build (and possibly an older GTK3 version), says its sending the event immediately instead of after binding.

**Comment by Consolatis** [comment]
I don't quite understand (I didn't read the relevant sfwbar code for this though) why the bar isn't requesting the correct output to begin with (e.g. `gtk_layer_set_monitor()`) if one is defined via the config. Maybe I am missing something here.

**Comment by LBCrion** [comment]
The bar launches by default on the current output, but can be moved to a
specific output if preferred. I would rather not introduce a requirement to
specify an output.

Of course one way to solve this would be start the bar on the first output
rather than the current output by default.

**Comment by Consolatis** [comment]
I think you can just not call `gtk_layer_set_monitor()` for bars that don't have `-m` / `SetMonitor` set and it should spawn on the focused output.

**Comment by keshto** [comment]
The current behavior is they spawn on the focused output without setting anything already.... The problem is when SetMonitor is used they dont span on the set monitor

**Comment by LBCrion** [comment]
I think I have a reasonable solution on this. We can avoid the race condition by querying/saving the  output when hiding the bar rather on init. @keshto - can you try the latest git please?

@Consolatis - bar output mapping and hide/show login is fairly complex, mainly due to dealing with cases such as outputs going away or coming back while the bar is hidden. In general, that bar tries to stay on the output on which it was launched or where it was assigned to via SetMonitor. If that output goes away and it wasn't set as static: in SetMonitor, the bar will jump to the primary (or first available) output.

**Comment by keshto** [comment]
Looks like it works as expected now. thanks!

One nit pick. When launching with no `setMonitor` the bar goes to the focused output which makes sense, but when launching with  `setMonitor` with an invalid output (non existent) the bar is launched on the first (laptop screen) even if it is launched from the external monitor. Shouldn't the behavior be same as without it being set?

**Comment by LBCrion** [comment]
When SetMonitor is called with a non-existent output, there is no way for sfwbar to know whether the output doesn't exist at all or just isn't currently connected. So sfwbar assigns this name as the bar's home output and since it doesn't currently exist, it follows the usual logic and moves the bar to a primary (or first) output. Then if the output with the given name is connected, the bar will move to it.

The alternative would be to ignore SetMonitor if the output with the given name doesn't exit, but this will break the scenario where you want the bar to move to an output if it appears. I.e. you run a bar with SetMonitor "DP-1" when external monitor isn't connected. Currently, the bar will move to the external monitor once it's connected, if we ignore SetMontitor, the move won't happen.
